### DISCUSSION TITLE
Searching for two words is not typo/plural resistant

### DISCUSSION THREAD

**Discussion by Boscop** [meilisearch/product#97] [Feedback & Feature Proposal]
**Describe the bug**
When I remove the `s` (so that the query is "moth flame"), this search result disappears (leading to zero results). Any idea why?

![image](https://user-images.githubusercontent.com/535593/116008487-2201cf00-a615-11eb-8379-10c91502e985.png)

(I'm using the online sandbox instance btw)

**To Reproduce**
1. Go to online sandbox
2. Enter "moths flame"
3. Remove the `s`
4. See behavior

**Expected behavior**
The search result should still appear.

---

Btw, a similar issue is when the query is "sky moth", this movie doesn't appear, but it should.

**Comment by CaroFG** [comment]
Hi! This is due to the fact that the typo tolerance rule doesn't allow typos on query words that are between 1 and 4 characters long. Since `moth` it's just 4 characters long, it isn't supposed to contain any typos. If you type `mothr flame` you'll get results. You can find more information about how typo tolerance works in [the dedicated section of the docs](https://docs.meilisearch.com/reference/under_the_hood/typotolerance.html#typo-tolerance-rules) :)

**Comment by Boscop** [comment]
I see, but it should still show up as search result, because both words (`moth` and `flame`) are substrings of the movie description!

**Comment by curquiza** [comment]
> I see, but it should still show up as search result, because both words (moth and flame) are substrings of the movie description!

What do you mean by `substrings of the movie description`?

You typed `moth` (4 characters) which is considered as a word with 1 typo compared to `moths`.
If you have typed only `moth` and not `moth flame`, you would have found this movie because you are in [prefix search](https://docs.meilisearch.com/reference/under_the_hood/prefix.html#prefix-search) for the word `moth`.
But since you have typed `moth` + another word (`flame`) you are not in prefix search anymore for the word `moth`: the word `moth` is considered as "over" and it contains 1 typo compared to `moths`. But, as @CaroFG said, typos on words containing 1 to 4 characters are not taken into account for the relevancy.
I hope this is clearer 🙂

**Comment by Boscop** [comment]
@curquiza I understand your reasoning but I'm saying as a human I would expect that search result to appear, because both words I entered are sub strings of the movie description.
And other search engines would show this movie as a search result.
Can you please consider changing your matching policy such that if all words of the query appear in the document, it appears as a search result? 
That would be way more intuitive, and then the user doesn't have to wonder if any documents are not caught by the query, that contain all its words.
Otherwise the user is left wondering about the potential documents that don't appear in search results.

**Comment by curquiza** [comment]
I completely understand the non-intuitive behavior of this search result. What is needed here is the plural handling more than just the typo tolerance, which is applied in your situation and which leads to a non-intuitive behavior for you.
However, plural handling is a complex feature to add, some plural forms are indeed not easy to handle (ex: cheval/chevaux in French) but we can consider it in the future 🙂

If I'm correct, we don't have this feature on our roadmap yet, but you can [submit this idea on it](https://roadmap.meilisearch.com/tabs/1-under-consideration) ("submit an idea" button at the top right). We really appreciate having the feedback of our users and your point of view on it would be really helpful 🙂

The current workaround is to add a plural/singular list of words as synonyms in MeiliSearch: https://docs.meilisearch.com/reference/features/settings.html#synonyms

**Reply by aradalvand** [reply]
You keep missing the obvious point and trying to "justify" this behavior... smh...
No, this is NOT about plural handling.

**Comment by Boscop** [comment]
In my use case it's not about plurals but about matching substrings. (It was just accidentally a plural in the example.)
My indexed data contains many typos, so I want to make the search more robust by not just matching whole words but substrings.

**Comment by eikooc** [comment]
@curquiza I think it also makes more sense to "stay in prefix search" when adding multiple words.

As you said earlier:
> But since you have typed moth + another word (flame) you are not in prefix search anymore for the word

I would suspect it would be more natural to _always_ be in prefix search mode, even for multiple words.
As someone searching I would get back a list of results by my first query.

```
query: "bat"
result: [{title: "Batman", year: "1984"}, {title: "Batman Begins", year: "2005"}]
```

And then if I wanted to drill down I would do:

```
query: "bat 20"
result: [{title: "Batman Begins", year: "2005"}]
```

There are even better and longer examples with company names and addresses.

**Reply by curquiza** [reply]
@meilisearch/product-team for the user feedback :)

**Reply by eikooc** [reply]
Did this comment mean I should post the feedback somewhere else or did you just tag them so they noticed? I still believe this feature would be awesome search UX

**Reply by macraig** [reply]
@eikooc there's no need to post the feedback somewhere else, although you can always submit or upvote the feature on our [public roadmap](https://roadmap.meilisearch.com/c/61-policy-for-prefix-search) if you like :)

> I still believe this feature would be awesome search UX

We agree it's a great feature to add! However, we're currently busy with other preceding priorities. We'll make sure to follow up on this discussion and on the roadmap I shared above once there's progress in this area.

**Reply by curquiza** [reply]
Also, I think the need is more related to https://github.com/meilisearch/product/discussions/640 rather than plural resistancy
